OpenRA.Mods.Common/Traits/BotModules/Squads/States/GroundStates.cs

```diff
@@ -68,57 +84,113 @@ public void Tick(Squad owner)
 
 	class GroundUnitsAttackMoveState : GroundStateBase, IState
 	{
+		const int MaxAttemptsToAdvance = 6;
+		const int MakeWayTicks = 2;
+
+		// Give tolerance for AI grouping team at start
+		int attemptsToAdvance = MaxAttemptsToAdvance * 3;
+		int makeWay = MakeWayTicks;
+		WPos lastPos = WPos.Zero;
+
 		public void Activate(Squad owner) { }
 
 		public void Tick(Squad owner)
 		{
+			// Basic check
 			if (!owner.IsValid)
 				return;
 
 			if (!owner.IsTargetValid)
 			{
-				var closestEnemy = FindClosestEnemy(owner);
-				if (closestEnemy != null)
-					owner.TargetActor = closestEnemy;
+				var targetActor = GetRandomPreferredTarget(owner);
+				if (targetActor != null)
+					owner.TargetActor = targetActor;
 				else
 				{
 					owner.FuzzyStateMachine.ChangeState(owner, new GroundUnitsFleeState(), true);
 					return;
 				}
 			}
 
-			var leader = owner.Units.ClosestTo(owner.TargetActor.CenterPosition);
+			// Initialize leader. Optimaze pathfinding by using leader.
+			var leader = owner.Units.FirstOrDefault();
 			if (leader == null)
 				return;
 
-			var ownUnits = owner.World.FindActorsInCircle(leader.CenterPosition, WDist.FromCells(owner.Units.Count) / 3)
-				.Where(a => a.Owner == owner.Units.First().Owner && owner.Units.Contains(a)).ToHashSet();
+			// 1. Scanning if there is enemy around and switch to GroundUnitsAttackState if any.
```

Comment:
Looking at this again, I don't see much need for the numbering. I suggest rewording this to
```
// Switch to GroundUnitsAttackState if we encounter enemy units.
// This makes sure that the entire squad will stay together and fight, instead of splitting
// into two groups based on whether they were close enough for AttackMove to acquire the target.
```

and then removing the numbering from 2 and 3.

Reply:
deleted.